adding complex support for distributed functions  and . fix #45760

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45879 adding complex support for distributed functions  and . fix #45760**

Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

bdhirsh committed:
adding complex support for distributed functions  and . fix #45760

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -216,10 +216,10 @@ def _build_tensor(size, value=None, dtype=torch.float):
+def _build_multidim_tensor(dim, dim_size, value=None, dtype=torch.float):
     if value is None:
         value = size
-    return torch.FloatTensor(size=[dim_size for _ in range(dim)]).fill_(value)
```
What's the use case for typed Tensor classes that are typed on their dtype, vs. the generic torch.Tensor class?

mrshenli replied:
not sure about the history here, but I think your solution is better.

bdhirsh commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1060,6 +1057,41 @@ def test_all_reduce_sum_cuda(self):
             )
 
+        @unittest.skipIf(BACKEND == "nccl", "Nccl does not support CPU tensors")
+        def test_all_reduce_sum_complex(self):
```
I only bothered adding complex tests for a single reduction op (`sum`) since it's unrelated to the actual op logic

bdhirsh replied:
update to that since the above comment isn't really true- added explicit tests that a complex-unsupported reduceOp like `Max` should error out properly :)

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"




[ghstack-poisoned]

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +935,32 @@ def all_reduce(tensor,
         None, if not async_op or if not part of the group
 
+    Example:
+        Tensors are all of dtype torch.int64.
```
double quote on all code 

```
``torch.int64``
``tensor = [[1, 1], [2, 2]]``
```

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +935,32 @@ def all_reduce(tensor,
+        rank 1 passes:
+            tensor = [[3+3i, 3+3i], [4+4i, 4+4i]]
+        both rank 0 and 1 get:
+            tensor = [[4+4i, 4+4i], [6+6i, 6+6i]]
```
could you please build the doc to verify that this renders correctly?

bdhirsh replied:
synced privately on this- I updated the doc to use python repl formatting for the examples and tested out building the doc:

![image](https://user-images.githubusercontent.com/16311747/95377462-9c087b80-08b0-11eb-88cf-09fa8dec64cd.png)

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +935,36 @@ def all_reduce(tensor,
+
+        >>> # Tensors are all of dtype torch.complex64.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([complex(1, 1), complex(2, 2)], dtype=torch.complex64) + 2 * complex(rank, rank)
```
`tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cdouble) + 2 * rank * (1+1j)` 

Unlike C++, Python can interpret the imaginary number `j`.  Also, maybe we should use `torch.complex128`  or `torch.cdouble` since above, we show an example of `torch.int64` and not `torch.int32`

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +935,36 @@ def all_reduce(tensor,
+        tensor([4, 6]) # Rank 0
+        tensor([4, 6]) # Rank 1
+
+        >>> # Tensors are all of dtype torch.complex64.
```
nit - All tensors below are of torch.complex64 dtype

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +935,36 @@ def all_reduce(tensor,
         None, if not async_op or if not part of the group
 
+    Examples:
+        >>> # Tensors are all of dtype torch.int64.
```
nit - All tensors below are of torch.int64 dtype

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1408,12 +1450,44 @@ def all_gather(tensor_list,
         None, if not async_op or if not part of the group
 
+    Examples:
+        >>> # Tensors are all of dtype torch.int64.
```
nit - All tensors below are of torch.int64 dtype

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1408,12 +1450,44 @@ def all_gather(tensor_list,
+        [tensor([1, 2]), tensor([3, 4])] # Rank 0
+        [tensor([1, 2]), tensor([3, 4])] # Rank 1
+
+        >>> # Tensors are all of dtype torch.complex64.
```
nit - All tensors below are of torch.complex64 dtype

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1408,12 +1450,44 @@ def all_gather(tensor_list,
+        >>> tensor_list = [torch.zero(2, dtype=torch.complex64) for _ in range(2)]
+        >>> tensor_list
+        [tensor([0.+0.j, 0.+0.j]), tensor([0.+0.j, 0.+0.j])] # Rank 0 and 1
+        >>> tensor = torch.tensor([complex(1, 1), complex(2, 2)], dtype=torch.complex64) + 2 * complex(rank, rank)
```
`tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cdouble) + 2 * rank * (1+1j)`

anjali411 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1198,33 +1230,37 @@ def test_sparse_all_reduce_sum_cuda(self):
                 [2, 3],
-                [1, 3]
+                [1, 3],
+                [torch.float, torch.float],
```
might be useful to explicitly mention in the documentation here: tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cdouble) + 2 * rank * (1+1j) that min, max are not supported for complex tensors.

bdhirsh replied:
agreed. I also added explicit error checking for that case

anjali411 replied:
sorry I messed up in copy pasting earlier. updated the comment with the link to the doc I was referring to

bdhirsh replied:
yep that's probably a better place to put it. added

gchanan reviewed (commented):
does complex support all of the reduce ops?  Like, I don't think we support max with complex and it doesn't seem like viewing it as a real will give you something that makes a ton of sense.

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

bdhirsh commented:
> does complex support all of the reduce ops? Like, I don't think we support max with complex and it doesn't seem like viewing it as a real will give you something that makes a ton of sense.

Agreed. I added checks so that we can error out accordingly- when the input tensor(s) are of a complex type, but the ReduceOp passed in does not support complex.

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -929,11 +947,39 @@ def all_reduce(tensor,
+
+        >>> # All tensors below are of torch.cdouble type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cdouble) + 2 * rank * (1+1j)
```
actually maybe it's best to let it be `torch.cfloat` because that's the default complex type if the default dtype is set to torch.float. This is consistent with the above example since default int dtype is `torch.long` or `torch.int64`.
If we want to use `torch.cdouble`, all the following prints of tensor, would look like:
`tensor([1.+1.j, 2.+2.j], dtype=torch.complex128)`

bdhirsh replied:
Makes sense. done

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1408,12 +1468,44 @@ def all_gather(tensor_list,
+
+        >>> # All tensors below are of torch.cfloat dtype.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor_list = [torch.zero(2, dtype=torch.float) for _ in range(2)]
```
nit - `torch.zero(2, dtype=torch.cfloat)`

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -43,6 +43,18 @@
+# We'd like calls to unsupported ops to error out accordingly,
+# rather than returning garbage values.
+def supports_complex(reduceOp: ReduceOp) -> bool:
+    if reduceOp == ReduceOp.MAX or reduceOp == ReduceOp.MIN or reduceOp == ReduceOp.PRODUCT:
+        return False
+    return True
```
is this same as:

```python
return True if reduceOp == ReduceOp.SUM else False?
```

bdhirsh replied:
cleaned this up a little to make it more pythonic.

mrshenli reviewed (approved):
Code LGTM! Please also get a stamp from @anjali411 

Besides, there are too many irrelevant test failures. Please rebase and rerun tests before landing.

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

anjali411 commented on torch/distributed/distributed_c10d.py:
```diff
@@ -44,6 +44,17 @@
+# We'd like calls to unsupported ops to error out accordingly,
+# rather than returning garbage values.
+def supports_complex(reduceOp: ReduceOp) -> bool:
+    denyList = [ReduceOp.MAX, ReduceOp.MIN, ReduceOp.PRODUCT]
```
this looks great!

anjali411 reviewed (approved):
LGTM overall. thanks Brian! My only other comment would be that we should add tests to ensure `BAND`, `BOR`, and `BXOR` work for complex.

@mrshenli will `torch.distributed.autograd` also work after this change?

bdhirsh committed:
Update on "adding complex support for distributed functions  and . fix #45760"


Differential Revision: [D24127949](https://our.internmc.facebook.com/intern/diff/D24127949)

[ghstack-poisoned]

gchanan commented on torch/distributed/distributed_c10d.py:
```diff
@@ -44,6 +44,17 @@
 except ImportError:
     _GLOO_AVAILABLE = False
 
+# Some reduce ops are not supported by complex numbers.
```
the way this comment is written reads like we allow calling them.

bdhirsh replied:
@gchanan quick fix here: https://github.com/pytorch/pytorch/pull/46599
